llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp

```diff
@@ -1835,6 +1835,40 @@ Instruction *InstCombinerImpl::foldSelectInstWithICmp(SelectInst &SI,
   if (Instruction *NewSel = foldSelectICmpEq(SI, ICI, *this))
     return NewSel;
 
+  // Very simple "GVN" on select arms if we have X == Y ? (binop X, Y) : ...
```

Comment:
Isn't it handled by `foldSelectValueEquivalence`?

Reply:
Yeah looks to be the case. Think the issue there is its a bit overly constrained. Think we can drop the `noundef` check if the resulting op is a constant.

Reply:
Update `foldSelectValueEquivalence` to work in this case.